crates/client/settlement/src/sync_state.rs

```diff
@@ -0,0 +1,207 @@
+use std::sync::Arc;
+
+use futures::StreamExt;
+use mp_block::Block as StarknetBlock;
+use mp_digest_log::find_starknet_block;
+use mp_hashers::HasherT;
+use mp_snos_output::{MessageL1ToL2, MessageL2ToL1, StarknetOsOutput};
+use mp_transactions::compute_hash::ComputeTransactionHash;
+use mp_transactions::Transaction;
+use pallet_starknet::runtime_api::StarknetRuntimeApi;
+use sc_client_api::BlockchainEvents;
+use sp_api::{HeaderT, ProvideRuntimeApi};
+use sp_arithmetic::traits::UniqueSaturatedInto;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::traits::Block as BlockT;
+use starknet_api::transaction::TransactionHash;
+
+use crate::{RetryStrategy, SettlementProvider, SettlementWorker, StarknetState, UnrecoverableError};
+
+impl<B, H, SC> SettlementWorker<B, H, SC>
+where
+    B: BlockT,
+    H: HasherT,
+    SC: ProvideRuntimeApi<B> + HeaderBackend<B> + BlockchainEvents<B>,
+    SC::Api: StarknetRuntimeApi<B>,
+{
+    /// A thread responsible for updating (progressing) Starknet state on the settlement layer.
+    /// For now we use a simplified setup without validity proofs & DA, but in the future
+    /// Starknet state contract will also validate state transition against the fact registry.
+    /// That means we will need to publish state diffs and STARK proof before state update.
+    pub async fn sync_state(
+        substrate_client: Arc<SC>,
+        settlement_provider: Box<dyn SettlementProvider>,
+        retry_strategy: Box<dyn RetryStrategy>,
+    ) {
+        loop {
+            match Self::sync_state_proc(substrate_client.as_ref(), settlement_provider.as_ref()).await {
+                Ok(()) => {
+                    return;
+                }
+                Err(err) => {
+                    log::error!("[settlement] {err}");
+                    match retry_strategy.can_retry(&err).await {
+                        true => {
+                            log::info!("[settlement] Retrying after error");
+                            continue;
+                        }
+                        false => {
+                            log::error!("[settlement] Unrecoverable error, exiting");
+                            return;
```

Comment:
Maybe the signature should be `-> !` and this should be a `panic!`. Wdyt?

Reply:
Wow, thanks, I didn't know about this type!